teps/0007-conditions-beta.md

```diff
@@ -0,0 +1,610 @@
+As discussed in the background section, `Conditions` manifest themselves as `Tasks`. We can phase out the separate `Conditions` CRD and eventually deprecating it. In place of `Conditions`, we can use `Tasks` produce a `Result` called `Skip` with string values `True` or `False` -- to indicate whether the `Task` it’s guarding will be executed or not. When a user provides another value, the `Task` will evaluate it as an error. Thus, we won’t have `Conditions` to migrate to beta and won’t have to maintain the separate `Conditions` CRD. Moreover, the features supported by `Tasks` become readily available to be used for guarded execution of `Tasks`. Using `Results` allows us to distinguish between when a `Task` fails and when the `Task` used as a `Guard` evaluates to `False`. 
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: Task
+metadata:
+  name: file-exists
+spec:
+  parameters:
+    - name: path
+  resources:
+    - name: workspace
+      type: git
+  results:
+    - name: skip
+      description: boolean indicating whether guarded Task should be skipped
+  steps:
+  - name: check-file-exists
+    image: alpine
+    script: |
+      if [ -f $(resources.workspace.path)/$(params.path) ]; then
+        echo true | tee /tekton/results/skip
+      else
+        echo false | tee /tekton/results/skip
+      fi
+---
+api: tekton.dev/v1alpha1
+kind: Pipeline
+metadata: 
+    name: generate-file
+spec:
+  workspaces:
+    - name: source-repo
+  params:
+    - name: path
+      default: 'README.md'
+  tasks: 
+    - name: create-file
+      taskRef: 
+        name: create-readme-file
+      workspaces:
+        - name: source
+          workspace: source-repo
+    - name: echo-file-exists 
+      when:
+        - name: check-file-exists
+          taskRef:
+            - name: file-exists
+          workspaces:
+            - name: source
+              workspace: source-repo
+      taskRef:
+        name: echo-file-exists
+```
+Initially, while still supporting `Conditions`, we can use `Tasks` used as `Guards`. Users can use either the `Conditions` list or `Guards` list, but not both of them. Over time, we will phase out the `Conditions` and support `Guards` only. 
+
+### Efficiency
+
+#### CelRun Custom Task
+
+If we implement guarded execution of `Tasks` is implemented using [Tasks that produce Skip Result](#tasks-that-produce-skip-result), we can then extend it to use [`CustomTasks`](https://docs.google.com/document/d/10nQSeIse7Ld4fLg4lhfgUmNKtewfaFNET3zlMdRnBuQ/edit#heading=h.nz0qjg4cmzp0) to build and experiment with using string expressions for simple `Guards`. In Triggers, we use [Common Expression Language](https://opensource.google/projects/cel) for filtering using a `CEL` interceptor. We can provide a `CelRun CustomTask`, so that we experiment with `CEL` without putting it into the Tekton API. After experimentation with `CelRun` `CustomTask` for a while, we will revisit whether we want to add it to the Tekton API.
```

Comment:
nice! if this gets merged as "implementable" maybe as a follow up we can at least create an issue about creating a CELRun controller, tho it will no longer block this :D

Reply:
sounds good 👍